src/core/utils/channel_manager.cpp

```diff
@@ -273,14 +267,247 @@ void ChannelManager::HandleStateChanged(uint32_t aFlags)
 
     mState = kStateIdle;
-    mTimer.Stop();
+    StartAutoSelectTimer();
 
     otLogInfoUtil(GetInstance(), "ChannelManager: Channel successfully changed to %d", mChannel);
 
 exit:
     return;
 }
 
+#if OPENTHREAD_ENABLE_CHANNEL_MONITOR
+
+/**
+ * This function randomly chooses a channel from a given channel mask.
+ *
+ * @param[in]  aMask  A channel mask.
+ *
+ * @returns A randomly chosen channel from the given mask, or `ChannelMask::kChannelIteratorFirst` if the mask is empty.
+ *
+ */
+static uint8_t ChooseRandomChannel(const Mac::ChannelMask &aMask)
+{
+    uint8_t channel     = Mac::ChannelMask::kChannelIteratorFirst;
+    uint8_t numChannels = 0;
+    uint8_t randomIndex;
+
+    VerifyOrExit(!aMask.IsEmpty());
+
+    while (aMask.GetNextChannel(channel) == OT_ERROR_NONE)
+    {
+        numChannels++;
+    }
+
+    randomIndex = Random::GetUint8InRange(0, numChannels);
+
+    channel = Mac::ChannelMask::kChannelIteratorFirst;
+    SuccessOrExit(aMask.GetNextChannel(channel));
+
+    while (randomIndex-- != 0)
+    {
+        SuccessOrExit(aMask.GetNextChannel(channel));
+    }
+
+exit:
+    return channel;
+}
+
+otError ChannelManager::FindBetterChannel(uint8_t &aNewChannel, uint16_t &aQuality)
+{
+    otError          error   = OT_ERROR_NONE;
+    ChannelMonitor & monitor = GetInstance().GetChannelMonitor();
+    Mac::ChannelMask favoredAndSupported;
+    Mac::ChannelMask favoredBest;
+    Mac::ChannelMask supportedBest;
+    uint16_t         favoredQuality;
+    uint16_t         supportedQuality;
+    char             string[Mac::ChannelMask::kInfoStringSize];
+
+    if (monitor.GetSampleCount() <= kMinChannelMonitorSampleCount)
+    {
+        otLogInfoUtil(GetInstance(), "ChannelManager: Too few samples (%d <= %d) to select channel",
+                      monitor.GetSampleCount(), kMinChannelMonitorSampleCount);
+        ExitNow(error = OT_ERROR_INVALID_STATE);
+    }
+
+    favoredAndSupported = mFavoredChannelMask;
+    favoredAndSupported.Intersect(mSupportedChannelMask);
+
+    favoredBest   = monitor.FindBestChannels(favoredAndSupported, favoredQuality);
+    supportedBest = monitor.FindBestChannels(mSupportedChannelMask, supportedQuality);
+
+    otLogInfoUtil(GetInstance(), "ChannelManager: Best favored %s, quality 0x%04x",
+                  favoredBest.ToString(string, sizeof(string)), favoredQuality);
+    otLogInfoUtil(GetInstance(), "ChannelManager: Best overall %s, quality 0x%04x",
+                  supportedBest.ToString(string, sizeof(string)), supportedQuality);
+
+    // Prefer favored channels unless there is no favored channel,
+    // or the quality of the best favored channel is worse than
+    // the best overall by at least `kQualityThresholdToSkipFavored`.
+
+    if (favoredBest.IsEmpty() || ((favoredQuality >= kQualityThresholdToSkipFavored) &&
+                                  (supportedQuality < favoredQuality - kQualityThresholdToSkipFavored)))
```

Comment:
Add some preprocessor checks to the config so that we do not need to worry about underflow here?